File: test/module.js
const expect = require('chai').expect,
  conf = require('./conf'),
  WhmcsError = require('../lib/whmcserror'),
  WhmcsResponse = require('../lib/whmcsresponse');

describe('Module "Module"', function () {

  it('should get the module queue', async function () {
    const res = await conf.whmcs.module.getModuleQueue();
    expect(res).to.be.an.instanceOf(WhmcsResponse);
    expect(res.getBody()).to.have.a.property('result').to.equal('success');
    expect(res.getBody()).to.have.a.property('queue').to.be.an('array');
  });

  it('should get the module configuration parameters', async function () {
    const opts = {
      moduleType: 'gateway',
      moduleName: 'paypal'
    };

    const res = await conf.whmcs.module.getModuleConfigurationParameters(opts);
    expect(res).to.be.an.instanceOf(WhmcsResponse);
    expect(res.getBody()).to.have.a.property('result').to.equal('success');
    expect(res.getBody()).to.have.a.property('parameters').to.be.an('array');
  });

  it('should activate a module', async function () {
    const opts = {
      moduleType: 'gateway',
      moduleName: 'paypal'
    };

    let res;

    try {
      res = await conf.whmcs.module.activateModule(opts);
      expect(res).to.be.an.instanceOf(WhmcsResponse);
      expect(res.getBody()).to.have.a.property('result').to.equal('success');
    } catch (e) {
      if (e instanceof WhmcsError) {
        const possibleErr = ['Failed to activate:', 'An unexpected error occurred:', 'Module activation not supported by module type.', 'Invalid module name provided.', 'Invalid module type provided. Supported module types include:'];
        expect(possibleErr.some(err => {
          return e.message.indexOf(err) > -1;
        })).to.be.true;
      } else {
        throw e;
      }
    }
  });

  it('should update module configuration parameters', async function () {
    const opts = {
      moduleType: 'gateway',
      moduleName: 'paypal'
    };

    const res = await conf.whmcs.module.updateModuleConfiguration(opts);
    expect(res).to.be.an.instanceOf(WhmcsResponse);
    expect(res.getBody()).to.have.a.property('result').to.equal('success');
  });

  it('should deactivate a module', async function () {
    const opts = {
      moduleType: 'gateway',
      moduleName: 'paypal',
      newGateway: 'paypal'
    };

    try {
      const res = await conf.whmcs.module.deactivateModule(opts);
      expect(res).to.be.an.instanceOf(WhmcsResponse);
      expect(res.getBody()).to.have.a.property('result').to.equal('success');
    } catch (e) {
      if (e instanceof WhmcsError) {
        expect(e.message).to.have.string('Module deactivation not supported by module type');
      } else {
        throw e;
      }
    }
  });

});

File: route/api/module.rb
namespace "/api/v1/module" do

  # GetModuleQueue
  get "/module/queue" do
    response = Whmcs.call("GetModuleQueue", { :relatedId => params[:relatedId].to_i,
                                              :serviceType => params[:serviceType].to_s,
                                              :moduleName => params[:moduleName].to_s,
                                              :moduleAction => params[:moduleAction].to_s,
                                              :since => params[:since].to_s })

    logger.info "[#{response[:result].upcase}] - #{response[:message]}."
    return json(response)
  end

  # ActivateModule
  get "/module/activate" do
    response = Whmcs.call("ActivateModule", { :moduleType => params[:moduleType].to_s,
                                              :moduleName => params[:moduleName].to_s,
                                              :parameters => params[:parameters].to_a })

    logger.info "[#{response[:result].upcase}] - #{response[:message]}."
    return json(response)
  end

  # DeactivateModule
  get "/module/deactivate" do
    response = Whmcs.call("ActivateModule", { :moduleType => params[:moduleType].to_s,
                                              :moduleName => params[:moduleName].to_s,
                                              :newGateway => params[:newGateway].to_s })

    logger.info "[#{response[:result].upcase}] - #{response[:message]}."
    return json(response)
  end

  # GetModuleConfigurationParameters
  get "/module/configuration" do
    response = Whmcs.call("GetModuleConfigurationParameters", { :moduleType => params[:moduleType].to_s,
                                                                :moduleName => params[:moduleName].to_s })

    logger.info "[#{response[:result].upcase}] - #{response[:message]}."
    return json(response)
  end

  # UpdateModuleConfiguration
  post "/module/configuration/update" do
    payload = JSON.parse(request.body.read.to_s)
    response = Whmcs.call("UpdateModuleConfiguration", { :moduleType => payload[:moduleType].to_s,
                                                         :moduleName => payload[:moduleName].to_s,
                                                         :parameters => payload[:parameters].to_a })

    logger.info "Module Configuration Updated:  #{response}"
    return json(response)
  end
end
